### ISSUE TITLE
Unexpected behaviour in `core:text/regex` regarding global flag

### ISSUE THREAD

**Issue by laytan** [odin-lang/Odin#4557]
Over the last few days (advent of code meaning new people or existing people trying the package and language out) we've had a few questions from people regarding unexpected behaviour in `core:text/regex`. To be clear with "unexpected" I mean different from most/all other implementations.

# Problems

## Patterns are anchored by default

Patterns seem to be anchored to the start of the given input by default, and the behaviour people expect from regex is that a pattern matches wherever in the input (which at the moment requires the .Global flag, more on that later).

For example, the following assertion fails:

```odin
rg, err := regex.create_by_user(`/world/`)
assert(err == nil)

_, ok := regex.match(rg, "hello world")
assert(ok) // Fails, what??
```

The expected behaviour is for it to match. And if one wants it anchored they would use `/^world/` to say so.

## The global flag is either unfortunately named, or wrong

The global flag and `/g` currently seem to mean "match wherever in the input" (which should be the default, see previous point). The expected behaviour is for `/g` in regex to mean "do not stop after the first match".

One would expect this:

```odin
rg, err := regex.create_by_user(`/(hello\d)/g`) 
assert(err == nil)

captures, ok := regex.match(rg, "hello0, hello1, hello2")
assert(ok)
fmt.println(captures) // ["hello0, hello1, hello2", "hello0", "hello1", "hello2"]
```

but that is not the behaviour.

# Suggestions

For the first point I think that the entire global flag should not exist and anchors should be done in the pattern, I believe that's how it works everywhere and is what people expect. If it is kept, it should at least be renamed so it doesn't conflict with what people expect the "global" or /g flag to do.

For the other point I would say either make the Global (/g) flag do what was expected or remove the flag entirely and outline that this is out of scope.

**Comment by laytan** [comment]
Pinging @Feoramund as the author of the package (and many more great things).

**Comment by Snikimonkd** [comment]
Hey, "new people" here :) Was trying to solve my aoc problem with regex for about 2 hours :(, then decided to watch tests for regex package as example, even wrote some by myself. I was expecting /g to work exactly as you said - find ALL matches in the string. Maybe my tests will be useful:
https://github.com/Snikimonkd/Odin/blob/regex_global_flag_issue/tests/core/text/regex/test_core_text_regex.odin#L977
https://github.com/Snikimonkd/Odin/blob/regex_global_flag_issue/tests/core/text/regex/test_core_text_regex.odin#L1096

**Comment by Feoramund** [comment]
I did document the behavior of `Global` both in the original PR (#3962) and `core:text/regex/common/common.odin`, though now I'm trying to remember why it is that way. (It's been a little while.)

I believe it has to do partly with being influenced by the Python API where `match` checks the whole string and `search` checks any substring of the string. That naming convention makes sense to me. There is also, of course, the articles of Russ Cox which dealt mainly with anchored matching by default.

Trying to refresh my memory, I popped open the virtual machine and had a look at the internals. I think the reason why `Global` exists is for performance. The API tries to do the most performant thing first (uses ASCII over Unicode, for example), and you have to opt-out with `Global`. That much I do remember: speed concerns influenced most of the default implementation.

---

Now, as for matching every possible match, such as you have shown here:
```
fmt.println(captures) // ["hello0, hello1, hello2", "hello0", "hello1", "hello2"]
```

This is impossible with the current virtual machine implementation, as the `Save` opcode places string indexes into slot arguments that are supplied at regex compile-time. It does not push these indexes into an unlimited list. This is explicitly a limit of the VM design and is hinted at by `regex_common.MAX_CAPTURE_GROUPS`.

The above expected output also does not make sense unless we are to have unbounded arrays of unions between strings and arrays. It gets messy. Each capture is a group, and each element in the capture result represents one group, so for `/(hello\d)/g` to make sense in this interpretation, you'd really want a result closer to `["hello0, hello1, hello2", ["hello0", "hello1", "hello2"]]` in order to make sense in the case where you have 4 distinct capture groups where you really do have separate groups for each `hello\d`. It makes the results between regex patterns consistent.

Setting that aside, I am not sure - without digging into it - if the design is capable of adapting to an unlimited number of matches as expected through this interpretation of `Global`. It may be possible, but the reliance on deterministic program counters (a thread at PC X should have the same state as any other thread at PC X) makes me feel like it could cause issues, most especially with dynamic memory allocation, if we're talking about unbounded arrays. It's something I'd have to think about again and do actual work on, given that it's been about 5 months since I last touched this package, but I'm already thoroughly tied up elsewhere.

---

Regarding what to do about this for now, I'd say if you have a situation where you're dependent on an unlimited number of matches (really, capture groups), it may be time to write your own parser where you can have the best performance possible with a design informed by your particular needs.

As for the `core:text/regex` API itself, I don't feel strongly right now about any particular resolution, having been thinking on this for the past hour or so, though I'm inclined to say it may be better to remove `Global` entirely if there is this confusion and encourage explicit `.*?` since it incurs a hidden cost if we make that the default behavior. Though, I did take another look at the VM just now, and I realize that using `Global` is the only way you get access to the specialized `Wait` opcodes, so if it's removed, a heuristic will have to be added to the optimizer in order to automatically insert those when any pattern begins with an unambiguous `.*?`

**Comment by laytan** [comment]
Could you not check if the pattern starts with ^ and decide based on it if it's bounded or not? I haven't seen this global or not behaviour before, I think Python is also an odd duck implementation having this.

If we can't do the behaviour of keeping on matching like expected with `/g`, it is unfortunate but not the end of the world, as long as we do something about the naming conflict between Global (/g) because seeing that has people expect it does support it, evident by the amount of questions about this over the past few days.

---

As for a resolution I would prefer:

If removing Global and determining anchors is possible, I would say remove it, and make /g an error, indicating that is unsupported, and maybe write something in the doc comment about stopping on the first match regardless of pattern.

If removing .Global is not possible, rename it to something else to avoid confusion.

---

PS: somebody on Discord wrote this to "emulate" the expected /g behaviour by rerunning the regex on the part of the string that hasn't been matched yet:
```odin
text := "m1m2m3"
pattern, pattern_err := regex.create("m\\d", { .Global })
assert (pattern_err == nil)

remainder := text
for capture in regex.match(pattern, remainder) {
    defer regex.destroy_capture(capture)

    fmt.println(capture.groups[0])
    remainder = remainder[capture.pos[0][1]:]
}
```

**Comment by Kelimion** [comment]
What we could do (sketch):
```odin
Match_Iterator :: struct {
    haystack: string,
    offset:   int,
    capture:  ^Capture,
    pos:      [common.MAX_CAPTURE_GROUP][2]int,
    groups:   [common.MAX_CAPTURE_GROUP]string,
}
```
First iteration, it sees `offset` is zero, allocates a `Capture` for internal use. Uses the static arrays to return `[][2]int`, `[]string` to return them to the user without them having to deallocate at the end of the iteration. On the last iteration, the iterator frees `^Capture` it used internally.

The Iterator can live on the stack, so no cleanup there either.
Usage would be something like `for match in regex.match_iterator(&it, haystack, pattern, options)`

But while that's nice and neat, this does unnecessary work (like compiling the regex) every iteration. A better approach might be:
```odin
it := regex.create_iterator(haystack, pattern, options)
defer regex.destroy(it)

for capture in regex.match_iterator(it) { ... }
```

Would use a similar Iterator struct but compile the pattern only once, making the whole thing more efficient, at the slight inconvenience of having to create and destroy the iterator. On the plus side, no having to destroy captures every iteration, as we slice them from the static arrays in the iterator.

As for `.Global`, it seems to function as "no implied `^` anchor", so I propose removing it and letting people use the start anchor `^` in their patterns if they insist that it matches at the start of the haystack.

**Comment by Kelimion** [comment]
I've created a PR with a match iterator, plus tests for it: #5008.

**Comment by Feoramund** [comment]
I figured out how I want to handle this and should have a PR up today.
